haskell.nix/library/flake.nix

```diff
@@ -0,0 +1,109 @@
+{
+  nixConfig = {
+    flake-registry = "https://github.com/serokell/flake-registry/raw/master/flake-registry.json";
+  };
+
+  inputs = {
+    flake-utils.url = "github:numtide/flake-utils";
+    flake-compat = {
+      flake = false;
+    };
+    haskell-nix = {
+      inputs.hackage.follows = "hackage";
+      inputs.stackage.follows = "stackage";
+    };
+    hackage = {
+      flake = false;
+    };
+    stackage = {
+      flake = false;
+    };
+  };
+
+  outputs = { self, nixpkgs, haskell-nix, hackage, stackage, serokell-nix, flake-compat, flake-utils, ... }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let
+        haskellPkgs = haskell-nix.legacyPackages."${system}";
+        pkgs = import nixpkgs {
+          inherit system;
+          overlays = [
+            serokell-nix.overlay
+          ];
+        };
+
+        lib = pkgs.lib;
+
+        hs-package-name = "pataq-package";
+
+        ghc-versions = [ "884" "8107" "902" ];
+
+        # invoke haskell.nix for each ghc version listed in ghc-versions
+        pkgs-per-ghc = lib.genAttrs (map (v: "ghc${v}") ghc-versions)
+          (ghc: haskellPkgs.haskell-nix.cabalProject {
+            src = haskellPkgs.haskell-nix.haskellLib.cleanGit {
+              name = hs-package-name;
+              src = ./.;
+            };
+            compiler-nix-name = ghc;
+
+            # haskell.nix configuration
+            modules = [{
+              packages.${hs-package-name} = {
+                ghcOptions = [
+                  # fail on warnings
+                  "-Werror"
+                  # disable optimisations, we don't need them if we don't package or deploy the executable
+                  "-O0"
+                ];
+              };
+
+            }];
+          });
+
+        # returns the list of all components for a package
+        get-package-components = pkg:
+          # library
+          lib.optional (pkg ? library) pkg.library
+          # haddock
+          ++ lib.optional (pkg ? library) pkg.library.haddock
+          # exes, tests and benchmarks
+          ++ lib.attrValues pkg.exes
+          ++ lib.attrValues pkg.tests
+          ++ lib.attrValues pkg.benchmarks;
+
+        # all components for each specified ghc version
+        build-all = lib.mapAttrs'
+          (ghc: pkg:
+            let components = get-package-components pkg.${hs-package-name}.components;
+            in lib.nameValuePair "${ghc}:build-all"
+              (pkgs.linkFarmFromDrvs "build-all" components)) pkgs-per-ghc;
```

Comment:
Hmm, now looking at this part once again, I'm starting to dislike the `linkFarmFromDrvs` approach.
I have a gut feeling that it's going to exit the build on the first build failure for some item from `components` without attempting to build the rest. This might be quite annoying because one won't be able to see all build issues in a single CI run.
However, I'm not sure how to approach this. For GA we can also use matrix approach, but this will produce huge build matrix (btw, can you check if it's possible to generate steps within a single job using build matrix?), which is probably fine as long as we have sufficient amount of runners. 
Unfortunately, I have no ideas at the moment how to approach this for GitLab CI and Buildkite since I cannot find if they support dynamic pipelines generation

Reply:
`--keep-going` sounds like a good option. I believe that in most cases builds aren't that large and it'll be possible to find an error in the logs